### PR TITLE
Create Network variant in reth_interfaces::Error

### LINKED ISSUE #936: Create `Network` variant in `reth_interfaces::Error`
I would also instead convert the error to an `RpcResult` rather than use `expect` here

I notice that we have helpers such as `ToRpcResult`, but this is only implemented for `reth_interfaces::Error`, which does not currently contain a variant for `Network`. This would require a larger refactor than in the scope of this PR

_Originally posted by @Rjected in https://github.com/paradigmxyz/reth/pull/923#discussion_r1081695626_
            
Creating a `Network` variant in `reth_interfaces::Error` would allow us to convert network errors to RPC errors more easily, using helpers such as `ToRpcResult`:
https://github.com/paradigmxyz/reth/blob/2a77e080809ec01371a7db1ff7327e7ed4b95519/crates/net/rpc/src/result.rs#L5-L36

### PR DESCRIPTION
**Pull request by literallymarvellous**
Resolves #936  

Draft

Created a Network Enum and have added it to ```reth_interfaces``` Error enum.

### CODE REVIEW

File: crates/interfaces/src/network.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use thiserror::Error;
+
+/// Network Errors
+#[allow(missing_docs)]
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum Error {
```
Comment (mattsse) [inline crates/interfaces/src/network.rs:6]: let's call this `NetworkError`

File: crates/net/network/src/network.rs [hunk 2]
```diff
@@ -227,14 +228,17 @@ impl PeersInfo for NetworkHandle {
...
         *self.inner.listener_address.lock()
     }
 
-    async fn network_status(&self) -> Result<NetworkStatus, Self::Error> {
-        let status = self.get_status().await?;
+    async fn network_status(&self) -> reth_interfaces::Result<NetworkStatus> {
+        let status = match self.get_status().await {
+            Ok(status) => status,
+            Err(_err) => return Err(reth_interfaces::network::Error::SenderDropped.into()),
+        };
```
Comment (mattsse) [inline crates/net/network/src/network.rs:241]: needs `From<SenderErr>` impl instead so we can keep using `?` operator

File: crates/net/network/src/network.rs [hunk 3]
```diff
@@ -227,14 +228,17 @@ impl PeersInfo for NetworkHandle {
 
 #[async_trait]
 impl NetworkInfo for NetworkHandle {
-    type Error = oneshot::error::RecvError;
+    type Error = Error;
```
Comment (mattsse) [inline crates/net/network/src/network.rs:231]: we should even get rid of this type and instead expect all functions to return `Resutl<T, NetworkError>` instead

File: crates/net/network/src/network.rs [hunk 4]
```diff
@@ -227,14 +228,17 @@ impl PeersInfo for NetworkHandle {
...
 
     fn local_addr(&self) -> SocketAddr {
         *self.inner.listener_address.lock()
     }
 
-    async fn network_status(&self) -> Result<NetworkStatus, Self::Error> {
-        let status = self.get_status().await?;
+    async fn network_status(&self) -> reth_interfaces::Result<NetworkStatus> {
+        let status = match self.get_status().await {
```
Comment (mattsse) [inline crates/net/network/src/network.rs:238]: NetworkError needs a variant for RecvErr,

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for picking this up,

left some suggestions,

please see `enum RequestError` for reference regard `From<Send|RecvErr>` impls

File: crates/net/network/src/network.rs (hunk 3 above)
Comment (literallymarvellous) [reply crates/net/network/src/network.rs:231]: Removing type Error here also means removing it from ```NetworkInfo```.
https://github.com/paradigmxyz/reth/blob/009d2056f95c6541b71eff386ee1477de4ebbf3b/crates/net/network-api/src/lib.rs#L17-L28

Trying to use the ```NetworkError``` enum but I can't get it to be accessible in ```network-api```, adding ```reth-interfaces = { path = "../../interfaces" }``` to its cargo.toml, throws a cyclic dependency error.

Comment (mattsse) [reply crates/net/network/src/network.rs:231]: I feel like NetworkError should be declared in network-api instead

and `interfaces` should consume it

File: crates/interfaces/src/network.rs [hunk 5]
```diff
@@ -0,0 +1,4 @@
+use reth_network_api::NetworkError;
+
+/// Error alias for Network error type
+pub type Error = NetworkError;
```
Comment (mattsse) [inline crates/interfaces/src/network.rs:4]: this file is no longer needed and can be removed.

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nit, otherwise, lgtm

File: crates/interfaces/src/network.rs (hunk 5 above)
Comment (literallymarvellous) [reply crates/interfaces/src/network.rs:4]: Alright

File: crates/net/network-api/src/lib.rs [hunk 6]
```diff
@@ -12,19 +12,38 @@
...
+    fn from(_: mpsc::error::SendError<T>) -> Self {
+        NetworkError::ChannelClosed
+    }
+}
+
+impl From<oneshot::error::RecvError> for NetworkError {
+    fn from(_: oneshot::error::RecvError) -> Self {
+        NetworkError::ChannelClosed
+    }
+}
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:37]: one last nit, sorry for being a bit pedantic, but would appreciate if you could move this to error.rs and  `pub use` from lib.rs

Comment (literallymarvellous) [reply crates/net/network-api/src/lib.rs:37]: Sure. Cleaner that way.

**Review by mattsse — APPROVED** [review]
ty

**Pull request merged**